File: server/src/models/Player.js
const mongoose = require("mongoose");
const autoIncrement = require("mongoose-auto-increment");

autoIncrement.initialize(mongoose.connection);

const availabilitySchema = mongoose.Schema({
  injuryStatus: { type: String, default: "" },
  injuryMessage: { type: String, default: "" },
});

const scoreSchema = mongoose.Schema({
  gameweekId: { type: Number },
  price: { type: Number },
  fantasyScore: { type: Number, default: 0 },
  minutesPlayed: { type: Number, default: 0 },
  goals: { type: Number, default: 0 },
  assists: { type: Number, default: 0 },
  cleanSheet: { type: Number, default: 0 },
  yellows: { type: Number, default: 0 },
  reds: { type: Number, default: 0 },
  penalitiesMissed: { type: Number, default: 0 },
  penalitiesSaved: { type: Number, default: 0 },
  saves: { type: Number, default: 0 },
  ownGoal: { type: Number, default: 0 },
  transfersIn: { type: Number, default: 0 },
  transfersOut: { type: Number, default: 0 },
  captainCount: { type: Number, default: 0 },
  viceCaptainCount: { type: Number, default: 0 },
  form: { type: Number, default: 0 },
});

const historySchema = mongoose.Schema({
  startingPrice: { type: Number },
  endingPrice: { type: Number },
  totalFanstasyScore: { type: Number },
  totalMinutesScore: { type: Number },
  totalGoals: { type: Number },
  totalAssists: { type: Number },
  totalCleanSheets: { type: Number },
  totalYellows: { type: Number },
  totalReds: { type: Number },
  totalPenalitiesMissed: { type: Number },
  totalPenalitiesSaved: { type: Number },
  totalSaves: { type: Number },
  totalOwnGoal: { type: Number },
  totalTransfersIn: { type: Number },
  totalTransfersOut: { type: Number },
  totalForm: { type: Number },
});

const playerSchema = mongoose.Schema({
  playerName: {
    unique: true,
    type: String,
    validate: /^[a-zA-Z,-,. ]*$/,
    minlength: 4,
    maxLength: 56,
    required: [
      true,
      "Custom Error - required Value *: Player Name is required.",
    ],
  },
  playerNameAmh: {
    // unique: true,
    type: String,
    // minlength: 4,
    // maxLength: 56,
    default: "",
    // required: [
    //   true,
    //   "Custom Error - required Value *: Player Name Amh is required.",
    // ],
  },
  eplTeamId: {
    type: String,
    minlength: 4,
    maxLength: 72,
    validate: /^[a-zA-Z,.,-,_ ]*$/,
    index: true,
    required: [
      true,
      "Custom Error - required Value *: Player Team is required.",
    ],
  },
  currentPrice: {
    type: Number,
    min: 3.5,
    required: [
      true,
      "Custom Error - required Value *: Player Price is required.",
    ],
  },
  position: {
    type: String,
    index: true,
    validate: /^[a-zA-Z,.,-,_ ]*$/,
    enum: {
      values: ["GK", "DEF", "MID", "ATT"],
      message:
        "Custom Error - Invalid ENUM Value *:{VALUE} is not allowed for game week status.",
    },
    required: [
      true,
      "Custom Error - required Value *: Player Position is required.",
    ],
  },
  availability: { type: availabilitySchema },

  playerImage: {
    type: String,
  },
  score: { type: [scoreSchema], default: [] },
  history: { type: [historySchema], default: [] },
});

playerSchema.plugin(autoIncrement.plugin, {
  model: "players",
  field: "playerId",
  startAt: 100000,
  incrementBy: 1,
});

module.exports = mongoose.model("players", playerSchema);

File: flutter-user-client/lib/domain/player/player.dart
import 'package:efpl/domain/player/value_objects.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'player.freezed.dart';

@freezed
abstract class Player with _$Player {
  const factory Player({
    required Name name,
    required Id playerId,
    required EplTeamId eplTeamId,
    required Position position,
    required Price currentPrice,
    required Availability availability,
    required Image image,
    @Default([]) List<Score> score,
    @Default([]) List<History> history,
  }) = _Player;
}

class Availability {
  InjuryStatus injuryStatus;
  InjuryMessage injuryMessage;

  Availability({required this.injuryStatus, required this.injuryMessage});
}

class Score {
  Gameweek gameweek;
  Price? price;
  FantasyScore? fantasyScore;
  MinutesPlayed? minutesPlayed;
  Goals? goals;
  Assists? assists;
  Cleansheet? cleansheet;
  Yellows? yellows;
  Reds? reds;
  PenalitiesMissed? penalitiesMissed;
  PenalitiesSaved? penalitiesSaved;
  Saves? saves;
  OwnGoal? ownGoal;
  TransfersIn? transfersIn;
  TransfersOut? transfersOut;
  Form? form;

  Score({
    required this.gameweek,
    this.price,
    this.fantasyScore,
    this.minutesPlayed,
    this.goals,
    this.assists,
    this.cleansheet,
    this.yellows,
    this.reds,
    this.penalitiesMissed,
    this.penalitiesSaved,
    this.saves,
    this.ownGoal,
    this.transfersIn,
    this.transfersOut,
    this.form,
  });
}

class History {
  Price? startingPrice;
  Price? endingPrice;
  FantasyScore? totalFantasyScore;
  MinutesPlayed? totalMinutesPlayed;
  Goals? totalGoals;
  Assists? totalAssists;
  Cleansheet? totalCleansheet;
  Yellows? totalYellows;
  Reds? totalReds;
  PenalitiesMissed? totalPenalitiesMissed;
  PenalitiesSaved? totalPenalitiesSaved;
  Saves? totalSaves;
  OwnGoal? totalOwnGoal;
  TransfersIn? totalTransfersIn;
  TransfersOut? totalTransfersOut;
  Form? totalForm;

  History({
    this.startingPrice,
    this.endingPrice,
    this.totalFantasyScore,
    this.totalMinutesPlayed,
    this.totalGoals,
    this.totalAssists,
    this.totalCleansheet,
    this.totalYellows,
    this.totalReds,
    this.totalPenalitiesMissed,
    this.totalPenalitiesSaved,
    this.totalSaves,
    this.totalOwnGoal,
    this.totalTransfersIn,
    this.totalTransfersOut,
    this.totalForm,
  });
}
